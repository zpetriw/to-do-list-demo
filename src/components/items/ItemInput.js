import React from 'react';
import { compose } from 'recompose';
import { withStyles } from 'material-ui/styles';
import PropTypes from 'prop-types';
import Button from 'material-ui/Button';
import TextField from 'material-ui/TextField';

const styles = theme => ({
    textField: {
        marginLeft: theme.spacing.unit,
        marginRight: theme.spacing.unit,
        width: 200,
    },
    button: {
        margin: theme.spacing.unit,
    }
});

export const ItemInput = ({classes, textLabel, textValue, buttonLabel, onChange, onClick, inputRef}) => {
    return (
        <div>
            <TextField
            required
            id="required"
            label={textLabel}
            className={classes.textField}
            margin="normal"
            onChange={onChange}
            value={textValue}
            inputRef={inputRef}
            />
            <Button variant="raised" className={classes.button} onClick={onClick}>
                {buttonLabel}
            </Button>
        </div>
    );
};

ItemInput.propTypes = {
    classes: PropTypes.object.isRequired,
    textLabel: PropTypes.string.isRequired,
    textValue: PropTypes.string.isRequired,
    buttonLabel: PropTypes.string.isRequired,
    onChange: PropTypes.func.isRequired,
    onClick: PropTypes.func.isRequired,
    inputRef: PropTypes.func,
};

export default compose(
    withStyles(styles)
)(ItemInput);