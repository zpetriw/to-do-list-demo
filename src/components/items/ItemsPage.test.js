import React from 'react';
import expect from 'expect';
import {shallow, render} from 'enzyme';
import configureMockStore from 'redux-mock-store';
import sinon from 'sinon';
import ItemInput from './ItemInput';
import Button from 'material-ui/Button';
import TextField from 'material-ui/TextField';
import ConnectedItemsPage, {ItemsPage} from './ItemsPage';
import * as itemActions from '../../actions/itemActions';
import renderer from 'react-test-renderer';

// Note: 
// toBe() means the two references must be for the same object.
// toEqual() does a deep equality check to see if two objects have the same keys.
// For primitives (int, string, ...) these functions will produce the same results.

describe('When ItemInput renders', () => {
    
    describe('its bare component', () => {
        
        let wrapper;
        let props;

        beforeEach( () => {
            props = {
                classes: {    
                    textField: "Text class",
                    button: "Button class",
                    root: "Root class"
                },
                textLabel: "Text label",
                buttonLabel: "Button label",
                items: [{title: "item1"}, {title: "item2"}],
                createItem: () => {}
            };
            
            wrapper = shallow(<ItemsPage {...props}/>);
        });
    
        it('should render correctly', () => {
            const tree = renderer
                .create(<ItemsPage {...props}/>)
                .toJSON();
            expect(tree).toMatchSnapshot();
        });

        it('should render <h1>, <h2>, and <ItemInput>', () => {
            expect(wrapper.find('h1').text()).toEqual('Items');
            expect(wrapper.find('h2').text()).toEqual('Add Item');
            expect(wrapper.find(ItemInput).length).toBe(1);
        });

        it('should render props.items to an array of itemRow', () => {
            expect(wrapper.find('div #items').children().length).toBe(2);
        });

        it('should pass component-bound props to ItemInput', () => {
            expect(wrapper.find(ItemInput).onClick).toBe();
            expect(wrapper.find(ItemInput).props().textValue).toEqual(wrapper.state().item.title);
            expect(wrapper.find(ItemInput).props().onChange).toEqual(wrapper.instance().onTitleChange);
            expect(wrapper.find(ItemInput).props().onClick).toEqual(wrapper.instance().onClickSave);
        });

    });
    
    describe('its withStyles(), Redux-connected component', () => {

        let wrapper;
        let store;
        let mockStore;
        let initialState = [{title: "item1"}, {title: "item2"}];
        let props;

        beforeEach(() => {
            mockStore = configureMockStore();
            store = mockStore({
                items: initialState
            });
            props = {
                textLabel: "Text label",
                buttonLabel: "Button label",
                items: [{title: "item1"}, {title: "item2"}],
                createItem: () => {}
            };

            wrapper = shallow(<ConnectedItemsPage store={store} {...props}/>).dive();
        });

        it('should render a WithStyles() className', () => {
            expect(wrapper.dive().props().className).toEqual(expect.stringContaining("ItemsPage"));
        });

        it('should map state.items to items', () => {
            expect(wrapper.dive().instance().props.items).toEqual(props.items);
        });

        it('should map action createItem', () => {
            expect(wrapper.dive().instance().props.createItem).not.toBeNull();
        });

    });

});