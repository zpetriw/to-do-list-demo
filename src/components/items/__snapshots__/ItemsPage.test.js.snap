// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`When ItemInput renders its bare component should render correctly 1`] = `
<div
  className="Root class"
>
  <h1>
    Items
  </h1>
  <div
    id="items"
  >
    <div>
      item1
    </div>
    <div>
      item2
    </div>
  </div>
  <h2>
    Add Item
  </h2>
  <div>
    <div
      aria-describedby={undefined}
      className="MuiFormControl-root-3 MuiFormControl-marginNormal-4 Component-textField-1"
      onBlur={[Function]}
      onFocus={[Function]}
    >
      <label
        className="MuiFormLabel-root-13 MuiInputLabel-root-7 MuiInputLabel-formControl-8 MuiInputLabel-animated-11"
        data-shrink={false}
        htmlFor="required"
      >
        Title
        <span
          className="MuiFormLabel-asterisk-17"
        >
          â€‰*
        </span>
      </label>
      <div
        className="MuiInput-root-19 MuiInput-formControl-20 MuiInput-underline-23"
      >
        <input
          aria-invalid={false}
          aria-required={true}
          autoComplete={undefined}
          autoFocus={undefined}
          className="MuiInput-input-27"
          defaultValue={undefined}
          disabled={false}
          id="required"
          name={undefined}
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          onKeyDown={undefined}
          onKeyUp={undefined}
          placeholder={undefined}
          readOnly={undefined}
          required={true}
          rows={undefined}
          type="text"
          value=""
        />
      </div>
    </div>
    <button
      className="MuiButtonBase-root-48 MuiButton-root-33 MuiButton-raised-38 Component-button-2"
      disabled={false}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex="0"
      type="button"
    >
      <span
        className="MuiButton-label-34"
      >
        Save
      </span>
      <span
        className="MuiTouchRipple-root-51"
      />
    </button>
  </div>
</div>
`;
